File: job_portal/src/app/core/services/content.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { of } from 'rxjs';

export interface ContentItem {
  id?: number;
  section: string;
  section_type: string;
  title: string;
  content?: string;
  image_url?: string;
  image_alt?: string;
  additional_data?: any;
  sort_order?: number;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface UserDashboardContent {
  hero: {
    title: string;
    subtitle: string;
    searchSuggestions: string;
    ctaButtonText: string;
  };
  welcome: {
    title: string;
    content: string;
    ctaButtonText: string;
    secondaryLinks: Array<{url: string, text: string}>;
  };
  howItWorks: {
    title: string;
    steps: Array<{
      icon: string;
      title: string;
      number: number;
      description: string;
    }>;
  };
  heroImage: {
    url: string;
    alt: string;
  };
}

interface ApiResponse<T> {
  success: boolean;
  data: T;
  message?: string;
}

@Injectable({
  providedIn: 'root'
})
export class ContentService {
  private apiUrl = 'http://localhost:3001/api/content';
  private dashboardContentSubject = new BehaviorSubject<UserDashboardContent | null>(null);
  public dashboardContent$ = this.dashboardContentSubject.asObservable();

  constructor(private http: HttpClient) {}

  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem('authToken');
    return new HttpHeaders({
      'Authorization': token ? `Bearer ${token}` : '',
      'Content-Type': 'application/json'
    });
  }

  // Public methods for user pages
  getUserDashboardContent(): Observable<UserDashboardContent> {
    return this.http.get<ApiResponse<UserDashboardContent>>(`${this.apiUrl}/public/user-dashboard`)
      .pipe(
        map(response => {
          this.dashboardContentSubject.next(response.data);
          return response.data;
        }),
        catchError(error => {
          console.error('Error fetching dashboard content:', error);
          // Return fallback content
          const fallbackContent: UserDashboardContent = {
            hero: {
              title: 'Find a job that suits your interest & skills.',
              subtitle: 'Explore thousands of job opportunities that match your skills and passions.',
              searchSuggestions: 'Designer, Programming, Digital Marketing, Video, Animation',
              ctaButtonText: 'Find Job'
            },
            welcome: {
              title: 'Welcome to Skillhunt',
              content: 'Create an account or sign in to see jobs that fit your requirements',
              ctaButtonText: 'Get Started',
              secondaryLinks: [
                { url: '/user/user-profile', text: 'Post your resume' },
                { url: '#', text: 'Post a job' }
              ]
            },
            howItWorks: {
              title: 'How Skillhunt Works',
              steps: [
                {
                  icon: 'fas fa-user-plus',
                  title: 'Create account',
                  number: 1,
                  description: 'Fill in all your details for setting up your profile visible to recruiters.'
                },
                {
                  icon: 'fas fa-upload',
                  title: 'Upload Resume',
                  number: 2,
                  description: 'Showcase your skills and experience with a standout CV.'
                },
                {
                  icon: 'fas fa-search',
                  title: 'Find suitable job',
                  number: 3,
                  description: 'Use smart filters to discover jobs tailored for you.'
                },
                {
                  icon: 'fas fa-paper-plane',
                  title: 'Apply Easily',
                  number: 4,
                  description: 'Send applications in one click and track your progress.'
                }
              ]
            },
            heroImage: {
              url: '/assets/person_searching_job.png',
              alt: 'Person searching job'
            }
          };
          this.dashboardContentSubject.next(fallbackContent);
          return of(fallbackContent);
        })
      );
  }

  getContentBySection(section: string): Observable<ContentItem> {
    return this.http.get<ApiResponse<ContentItem>>(`${this.apiUrl}/public/section/${section}`)
      .pipe(
        map(response => response.data),
        catchError(error => {
          console.error('Error fetching content by section:', error);
          return of({
            section,
            section_type: 'hero',
            title: 'Content not available',
            content: 'This content is currently being updated.'
          } as ContentItem);
        })
      );
  }

  getContentBySectionType(sectionType: string): Observable<ContentItem[]> {
    return this.http.get<ApiResponse<ContentItem[]>>(`${this.apiUrl}/public/section-type/${sectionType}`)
      .pipe(
        map(response => response.data),
        catchError(error => {
          console.error('Error fetching content by section type:', error);
          return of([]);
        })
      );
  }

  // Admin methods (require authentication)
  getAllContent(): Observable<any> {
    const headers = this.getAuthHeaders();
    return this.http.get<any>(`${this.apiUrl}/admin`, { headers })
      .pipe(
        catchError(error => {
          console.error('Error fetching all content:', error);
          // If unauthorized, return error with specific message
          if (error.status === 401) {
            return of({
              success: false,
              error: 'Authentication required. Please log in as admin.',
              data: []
            });
          }
          return of({
            success: false,
            error: error.message || 'Failed to load content',
            data: []
          });
        })
      );
  }

  createContent(content: Partial<ContentItem>): Observable<any> {
    const headers = this.getAuthHeaders();
    return this.http.post<any>(`${this.apiUrl}/admin`, content, { headers })
      .pipe(
        catchError(error => {
          console.error('Error creating content:', error);
          return of({
            success: false,
            error: error.message
          });
        })
      );
  }

  updateContent(id: number, content: Partial<ContentItem>): Observable<any> {
    const headers = this.getAuthHeaders();
    return this.http.put<any>(`${this.apiUrl}/admin/${id}`, content, { headers })
      .pipe(
        catchError(error => {
          console.error('Error updating content:', error);
          return of({
            success: false,
            error: error.message
          });
        })
      );
  }

  deleteContent(id: number): Observable<any> {
    const headers = this.getAuthHeaders();
    return this.http.delete<ApiResponse<any>>(`${this.apiUrl}/admin/${id}`, { headers })
      .pipe(
        map(response => response),
        catchError(error => {
          console.error('Error deleting content:', error);
          throw error;
        })
      );
  }

  // Helper method to refresh dashboard content cache
  refreshDashboardContent(): void {
    this.getUserDashboardContent().subscribe();
  }
}

File: job_portal_fastapi/app/services/content_service.py
from sqlalchemy.orm import Session
from app.db.models.content.content_model import SiteContent
from app.db.schemas.content.content_schema import SiteContentCreate, SiteContentUpdate
from fastapi import HTTPException

def get_all_content(db: Session):
    return db.query(SiteContent).order_by(SiteContent.section_type, SiteContent.sort_order, SiteContent.section).all()

def get_content_by_section(db: Session, section: str):
    return db.query(SiteContent).filter(SiteContent.section == section, SiteContent.is_active == True).order_by(SiteContent.sort_order).first()

def get_content_by_section_type(db: Session, section_type: str):
    return db.query(SiteContent).filter(SiteContent.section_type == section_type, SiteContent.is_active == True).order_by(SiteContent.sort_order, SiteContent.section).all()

def create_content(db: Session, content: SiteContentCreate):
    db_content = db.query(SiteContent).filter(SiteContent.section == content.section).first()
    if db_content:
        raise HTTPException(status_code=400, detail="Content section already exists")
    db_content = SiteContent(**content.dict())
    db.add(db_content)
    db.commit()
    db.refresh(db_content)
    return db_content

def update_content(db: Session, content_id: int, content_update: SiteContentUpdate):
    db_content = db.query(SiteContent).filter(SiteContent.id == content_id).first()
    if not db_content:
        raise HTTPException(status_code=404, detail="Content not found")

    update_data = content_update.dict(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_content, key, value)

    db.add(db_content)
    db.commit()
    db.refresh(db_content)
    return db_content

def delete_content(db: Session, content_id: int):
    db_content = db.query(SiteContent).filter(SiteContent.id == content_id).first()
    if not db_content:
        raise HTTPException(status_code=404, detail="Content not found")
    db.delete(db_content)
    db.commit()
    return {"message": "Content deleted successfully"}

def get_user_dashboard_content(db: Session):
    sections = ['hero', 'welcome', 'how_it_works']
    content_items = db.query(SiteContent).filter(SiteContent.section.in_(sections), SiteContent.is_active == True).order_by(SiteContent.sort_order).all()

    dashboard_content = {
      'hero': {'title': '', 'subtitle': '', 'searchSuggestions': '', 'ctaButtonText': 'Find Job'},
      'welcome': {'title': '', 'content': '', 'ctaButtonText': 'Get Started', 'secondaryLinks': []},
      'howItWorks': {'title': '', 'steps': []},
      'heroImage': {'url': '/assets/person_searching_job.png', 'alt': 'Person searching job'}
    }

    for item in content_items:
        additional_data = item.additional_data or {}
        if item.section == 'hero':
            dashboard_content['hero']['title'] = item.title
            dashboard_content['hero']['subtitle'] = item.content
            dashboard_content['hero']['searchSuggestions'] = additional_data.get('searchSuggestions', '')
            dashboard_content['hero']['ctaButtonText'] = additional_data.get('ctaButtonText', 'Find Job')
            if 'imageUrl' in additional_data:
                dashboard_content['heroImage']['url'] = additional_data['imageUrl']
            if 'imageAlt' in additional_data:
                dashboard_content['heroImage']['alt'] = additional_data['imageAlt']
        elif item.section == 'welcome':
            dashboard_content['welcome']['title'] = item.title
            dashboard_content['welcome']['content'] = item.content
            dashboard_content['welcome']['ctaButtonText'] = additional_data.get('ctaButtonText', 'Get Started')
            dashboard_content['welcome']['secondaryLinks'] = additional_data.get('secondaryLinks', [])
        elif item.section == 'how_it_works':
            dashboard_content['howItWorks']['title'] = item.title
            dashboard_content['howItWorks']['steps'] = additional_data.get('steps', [])

    return dashboard_content
